.github/workflows/main.yml

```diff
@@ -1,193 +1,135 @@
-# This is a basic workflow to help you get started with Actions
+name: XGBoost CI
 
-name: XGBoost-CI
-
-# Controls when the action will run. Triggers the workflow on push or pull request
-# events but only for the master branch
 on: [push, pull_request]
 
 permissions:
-  contents: read # to fetch code (actions/checkout)
+  contents: read  # to fetch code (actions/checkout)
 
 concurrency:
   group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
   cancel-in-progress: true
 
-# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+env:
+  BRANCH_NAME: >-
+    ${{ github.event.pull_request.number && 'PR-' }}${{ github.event.pull_request.number || github.ref_name }}
+  USE_DOCKER_CACHE: 1
+
 jobs:
-  gtest-cpu:
-    name: Test Google C++ test (CPU)
-    runs-on: ${{ matrix.os }}
+  build-containers:
+    name: Build CI containers (${{ matrix.container_id }})
+    runs-on:
+      - runs-on
+      - runner=${{ matrix.runner }}
+      - run-id=${{ github.run_id }}
+      - tag=main-build-containers-${{ matrix.container_id }}
     strategy:
-      fail-fast: false
       matrix:
-        os: [macos-12]
+        container_id:
+          - xgb-ci.gpu_build_rockylinux8
+          - xgb-ci.gpu
+          - xgb-ci.manylinux_2_28_x86_64
```

Comment:
> Is it necessary for correctness that all these images be rebuilt + re-pushed on every single commit in every PR?

The image is only rebuilt if the Dockerfile is updated in the pull request. All built images are uploaded to [Amazon Elastic Container Registry](https://aws.amazon.com/ecr/), and `docker_build.sh` will re-use the built images whenever possible. In effect, we are using ECR as a caching layer for images.

Reply:
> I think you should consider instead setting up the image builds in a separate repo in the dmlc organization, triggering them on a schedule, and then just pulling them here.

Doing is this way is definitely simpler. One disadvantage of this method is that we won't be able to test changes to Dockerfiles in the context of other Pytests, gtests, etc. How does RAPIDS test changes to images to ensure that they don't break tests?

Reply:
> *The image is only rebuilt if the Dockerfile is updated in the pull request.*

Where is that implemented, can you point me to it? Because it looks to me like `docker_build.sh` is always invoked on every run, and that it always runs `docker_build.py`, and that `docker_build.py` always runs `docker build`.

> *we won't be able to test changes to Dockerfiles in the context of other Pytests, gtests, etc. How does RAPIDS test changes to images to ensure that they don't break tests?*

Testing docker changes together with changes to the tests or other code here would look like this:

1. push new images from the repo where you build/publish images
2. on your PR branch here, change the tags to the new images you pushed
3. iterate until you're happy that everything is working together
4. merge the image-building PR
5. revert the testing-only tag changes here
6. merge the PR here

This is how @StrikerRUS and I have been managing CI images for LightGBM for a couple years, and it's worked pretty well for us (ref: https://github.com/guolinke/lightgbm-ci-docker).

In RAPIDS, we often test stuff like this by just making the changes in https://github.com/rapidsai/ci-imgs and then seeing what happens (knowing that reverting is as easy as just putting up a new revert PR in `ci-imgs` and merging it).

Occasionally, we follow a process similar to the one I described above, with the extra layer of indirection that "change the tags" means "put up a new branch in https://github.com/rapidsai/shared-workflows and switch GitHub Actions in PRs to pull from that branch".

Reply:
> Where is that implemented, can you point me to it? Because it looks to me like docker_build.sh is always invoked on every run, and that it always runs docker_build.py, and that docker_build.py always runs docker build.

Inside `docker_run.sh`, there is a block of code that first pulls a cached image from ECR. Then `docker build` is invoked with `--cache-from` argument. If no change was made to the Dockerfile, `docker build` becomes a no-op.

Reply:
Oh ok. So it is still doing a build, but you're just expecting it to hit the cache for every layer.

I wouldn't have expected stuff like this to be very cache-friendly:

```shell
dnf install -y tar unzip wget xz git which ninja-build
```

But it sure does look like `docker build` is hitting the cache for them:

```text
#9 [2/8] RUN     curl -fsSL https://developer.download.nvidia.com/compute/cuda/repos/rhel7/x86_64/D42D0685.pub | sed '/^Version/d'         > /etc/pki/rpm-gpg/RPM-GPG-KEY-NVIDIA &&     dnf -y update &&     dnf -y install dnf-plugins-core &&     dnf config-manager --set-enabled powertools &&     dnf install -y tar unzip wget xz git which ninja-build gcc-toolset-10-gcc gcc-toolset-10-binutils gcc-toolset-10-gcc-c++ &&     wget -nv -O conda.sh https://github.com/conda-forge/miniforge/releases/download/24.3.0-0/Miniforge3-24.3.0-0-Linux-x86_64.sh &&     bash conda.sh -b -p /opt/miniforge &&     /opt/miniforge/bin/python -m pip install awscli &&     wget -nv -nc https://cmake.org/files/v3.29/cmake-3.29.5-linux-x86_64.sh --no-check-certificate &&     bash cmake-3.29.5-linux-x86_64.sh --skip-license --prefix=/usr
#9 CACHE
```

([build link](https://github.com/dmlc/xgboost/actions/runs/12191835983/job/34013680923?pr=11046#step:5:75))

So I need to read up more on how caching works with docker BuildKit.

But even in that case, it DOES look like every single commit will try to push to ECR. Scroll further down in that same log, you'll see a tag and attempted push:

```text
docker tag xgb-ci.gpu_build_rockylinux8 492475357299.dkr.ecr.us-west-2.amazonaws.com/xgb-ci.gpu_build_rockylinux8:PR-11046
aws ecr create-repository --repository-name xgb-ci.gpu_build_rockylinux8 --region us-west-2

An error occurred (RepositoryAlreadyExistsException) when calling the CreateRepository operation: The repository with name 'xgb-ci.gpu_build_rockylinux8' already exists in the registry with id '4924753572[99](https://github.com/dmlc/xgboost/actions/runs/12191835983/job/34013680923?pr=11046#step:5:100)'
docker push --quiet 492475357299.dkr.ecr.us-west-2.amazonaws.com/xgb-ci.gpu_build_rockylinux8:PR-11046
492475357299.dkr.ecr.us-west-2.amazonaws.com/xgb-ci.gpu_build_rockylinux8:PR-11046
```

I still think it'd be better to not do that on every commit to every PR.

Reply:
Got it. I will go ahead and adopt your proposal, as follows:

* The container definitions will be moved to `dmlc/xgboost-devops`. In this repo:
  - The CI will build a new container each time a new commit is added to the `main` branch. Each container will be tagged with the branch name, e.g. `xgb-ci.gpu:main`.
  - Developers can open a new PR to `xgboost-devops` to trigger the build of a new container. The new container will be pushed to the ECR so that it's available to be used in a pull request for XGBoost. The new container will have tag of form `[container-name]:PR-[number]`.
* `docker_run.py` will now pull the container (if needed) and run it.
* `docker_build.py` will be moved to `xgboost-devops`.
* `docker_build.sh` will be removed.

@jameslamb What do you think?

Reply:
One more reason to adopt the new design: we've had issues in the past where Docker produces a false cache hit in the Conda install step, so that we'd end up with an old version of RAPIDS from the `rapidsai-nightly` channel. Using the new design, we will be able to re-build containers at at a scheduled interval and catch issues with nightly versions of Python packages in time.

Reply:
Thanks so much for considering it! I support the structure you described above, I think it'll be really helpful.

Reply:
Done. I've moved all the container building to https://github.com/dmlc/xgboost-devops/tree/main/containers and setup a separate CI pipeline to build the containers.